libp2p/switch.nim

```diff
@@ -256,21 +258,29 @@ proc start*(s: Switch) {.async, gcsafe.} =
       try:
         await t.start(addrs)
       except TransportListenError as e:
-        let err = s.listenError(e)
+        let err = await s.listenError(t, e)
         if not err.isNil:
           raise err
       s.acceptFuts.add(s.accept(t))
       s.peerInfo.addrs &= t.addrs
 
   debug "Started libp2p node", peer = s.peerInfo
 
-proc newSwitchListenError*(parent: ref TransportListenError = nil): ref SwitchListenError =
-  (ref SwitchListenError)(msg: "Failed to start one transport", parent: parent)
+proc newSwitchListenError*(
+    t: Transport,
+    parent: ref TransportListenError = nil): ref SwitchListenError =
+
+  (ref SwitchListenError)(msg: "Failed to start one transport",
+                          transport: t,
+                          parent: parent)
 
 const ListenErrorDefault =
```

Comment:
I'm not sure if it makes sense to provide a switch level callback at this point. If every transport is already given the chance to delegate error handling to the user, it's probably better to just asume that if a transport raised after the user inspected the error and decided to propagate it, there is no reason in keeping with the startup sequence. In other words, the switch will always abort if one of the transports failed, but the transport can decide wether the error is fatal or not. 

That way we eliminate the complexity involved in having two sets of callbacks at different levels, the added bonus is that transports can still handle errors individually on a per transport basis, with separate callbacks (or the same one if desired).

IMO, this would be a lot cleaner overall.

Reply:
I definitely agree this will simplify things quite a lot. Will update.